[Cocoa] Stop responsiveness checks on suspended web content processes

salinas-miguel opened the pull request:
#### 1f3dcbd1984c9e5548efef4d638f87d4dbfc3c66
<pre>
[Cocoa] Stop responsiveness checks on suspended web content processes
<a href="https://bugs.webkit.org/show_bug.cgi?id=253544">https://bugs.webkit.org/show_bug.cgi?id=253544</a>
rdar://problem/106388234

Reviewed by Chris Dumez.

We shouldn&apos;t be checking if suspended processes are responsive

* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::sendPrepareToSuspend):
(WebKit::WebProcessProxy::sendProcessDidResume):

Canonical link: <a href="https://commits.webkit.org/261382@main">https://commits.webkit.org/261382@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1680b5164931c8f97d8a1361035732e289420783

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/111541 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/20676 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/135 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/87/builds/3311 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/120309 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/22036 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/11765 "Build is in progress. Recent messages:") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/87/builds/3311 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/117304 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/16359 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/99494 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/104322 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/98315 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/77 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/45095 "") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/13167 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/75 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/86893 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/13665 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/85/builds/11765 "Build is in progress. Recent messages:") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/19110 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52058 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7924 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/15641 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

cdumez reviewed (commented):
I don't think this stops something from restarting the responsiveness timer after you've stopped it. Have you considered updating AuxiliaryProcessProxy::mayBecomeUnresponsive() instead to return false when a process is considered to be suspended?

cdumez commented on Source/WebKit/UIProcess/ProcessThrottler.h:
```diff
@@ -143,6 +143,7 @@ class ProcessThrottler : public CanMakeWeakPtr<ProcessThrottler> {
     void setAllowsActivities(bool);
     void setShouldTakeSuspendedAssertion(bool);
     void delaySuspension();
+    bool isSuspended() const { return !m_assertion; }
```
Patch looks good but can you please confirm that `isSuspended()` will not return true unconditionally for non-Cocoa ports? We wouldn't want to disable responsiveness checks on GTK port for example.

cdumez commented on Source/WebKit/UIProcess/ProcessThrottler.h:
```diff
@@ -143,6 +143,14 @@ class ProcessThrottler : public CanMakeWeakPtr<ProcessThrottler> {
     void delaySuspension();
+    bool isSuspended() const
+    {
+#if PLATFORM(COCOA)
```
I wish we wouldn't add build checks in this cross-platform code. I think we should make sure that other ports construct ProcessAssertions when needed. ProcessAssertion is cross-platform (though is a no-op on non-Cocoa ports at the moment).

I suggest we double check that GTK for example construct ProcessAssertions and ping a GTK port engineer for extra validation (E.g. @mcatanzaro).

salinas-miguel replied:
We only drop all assertions on Cocoa so `isSuspended()` should always return false on other platforms. I double checked that GTK is able to construct ProcessAssertions. I agree extra validation would be great. I'll ping @mcatanzaro

cdumez reviewed (approved):
LGTM without build-time check.

salinas-miguel committed:
[Cocoa] Stop responsiveness checks on suspended web content processes
https://bugs.webkit.org/show_bug.cgi?id=253544
rdar://problem/106388234

Reviewed by Chris Dumez.

We shouldn't be checking if suspended processes are responsive

* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::sendPrepareToSuspend):
(WebKit::WebProcessProxy::sendProcessDidResume):

Canonical link: https://commits.webkit.org/261382@main
